iothub_client/inc/iothub_module_client_ll.h

```diff
@@ -284,6 +284,22 @@ extern "C"
     * @return    IOTHUB_CLIENT_OK upon success or an error code upon failure.
     */
      MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubModuleClient_LL_SendReportedState, IOTHUB_MODULE_CLIENT_LL_HANDLE, iotHubModuleClientHandle, const unsigned char*, reportedState, size_t, size, IOTHUB_CLIENT_REPORTED_STATE_CALLBACK, reportedStateCallback, void*, userContextCallback);
+
+     /**
+     * @brief	This API enabled the device to request the full module twin (with all the desired and reported properties) on demand.
+     *
+     * @param	iotHubModuleClientHandle	The handle created by a call to the create function.
+     * @param	deviceTwinCallback	        The callback specified by the module client to receive the Twin document.
+     * @param	userContextCallback		    User specified context that will be provided to the
+     * 									    callback. This can be @c NULL.
+     *
+     *            @b NOTE: The application behavior is undefined if the user calls
+     *            the ::IoTHubModuleClient_LL_Destroy function from within any callback.
+     *
+     * @return	IOTHUB_CLIENT_OK upon success or an error code upon failure.
+     */
+     MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubModuleClient_LL_GetDeviceTwinAsync, IOTHUB_MODULE_CLIENT_LL_HANDLE, iotHubModuleClientHandle, IOTHUB_CLIENT_DEVICE_TWIN_CALLBACK, deviceTwinCallback, void*, userContextCallback);
```

Comment:
>DeviceTwin [](start = 72, length = 10)

This API should be called GetModuleTwinAsync. Also, I wanted to understand why the callback couldn't its own signature as IOTHUB_CLIENT_MOBULE_TWIN_CALLBACK?

Reply:
It probably should.  We shouldn't mix Device and Modules together.

---
In reply to: [242065498](https://github.com/Azure/azure-iot-sdk-c/pull/771#discussion_r242065498) [](ancestors = 242065498)

Reply:
Regarding the callback, we use the same callback all the way down the stack, at the transport level. Creating a separate callback definition will create duplication, so I am not sure it's advisable.
And renaming is not possible, because it would break the contract with the customers.